### PR TITLE
One NodeShared per process

### PR DESCRIPTION
**Pull request by joxoby**
https://github.com/orgs/ignitionrobotics/projects/3#card-38255453

### CODE REVIEW

File: src/NodeShared.cc [hunk 1]
```diff
@@ -139,12 +138,17 @@ NodeShared *NodeShared::Instance()
   // application exits. We may want to consider a more elegant solution in
   // the future. The zsys_shutdown() function in the czmq library may be able
   // to provide some inspiration for solving this more cleanly.
-  static NodeShared *instance = new NodeShared();
+  static NodeShared *instance = nullptr;
+
+  // Create a new instance of NodeShared if the process has changed
+  // (maybe after fork?) so the ZMQ context is not shared between different
+  // processes.
+  if (instance == nullptr || instance->pUuid != Uuid().ToString())
```
Comment (joxoby) [inline src/NodeShared.cc:146]: We may want to make this `Uuid` comparison faster

**Comment by caguero** [comment]
Thanks for the contribution. I've been doing some testing and I have two major concerns:

**Comment by joxoby** [comment]
Hi Carlos,
1. You're right, it was a confusion of mine regarding `Uuid`. What I meant to codify was something like
```
if (instance == nullptr || instance->pid != getpid())
```
2. We can easily add some `lock_guards`. WDYT?

**Comment by caguero** [comment]
> instance == nullptr || instance->pid != getpid()

Do you mind to update this pull request with a working prototype? It'll help to wrap my head around it. Thanks!

**Comment by joxoby** [comment]
With the last change, the `ignition-transport` test are passing but it does cause problems with `ign-gazebo`. Will update.

**Comment by caguero** [comment]
I'll decline this pull request for now until there's a better patch that keeps all tests and examples working. Feel free to reopen if that's the case.

**Pull request closed without merging**